### PR TITLE
Don't allocate DepNode if anonymous

### PR DESCRIPTION
**Pull request by arora-aman**
#45408

### CODE REVIEW

File: compiler/rustc_query_system/src/dep_graph/graph.rs [hunk 1]
```diff
@@ -260,6 +260,13 @@ impl<K: DepKind> DepGraph<K> {
 
             let print_status = cfg!(debug_assertions) && cx.debug_dep_tasks();
 
+            if matches!(dep_node_index, None) {
+                return (result, self.next_virtual_depnode_index());
+            }
+
+            // Handled None case above, safe to unwrap here.
+            let dep_node_index = dep_node_index.unwrap();
```
Comment (bjorn3) [inline compiler/rustc_query_system/src/dep_graph/graph.rs:268]: You can use
``rust
let dep_node_index = if let Some(dep_node_index) = dep_node_index {
    dep_node_index
} else {
    return ...
};
```

Comment (arora-aman) [reply compiler/rustc_query_system/src/dep_graph/graph.rs:268]: fair, I'll update once the perf run goes through

**Comment by bjorn3** [comment]
Up to 0.4% slowdown.

**Comment by crlf0710** [comment]
Triage: Is this ready for review? There's still [WIP] in the PR title.

File: compiler/rustc_middle/src/dep_graph/mod.rs [hunk 2]
```diff
@@ -85,6 +85,10 @@ impl rustc_query_system::dep_graph::DepKind for DepKind {
     fn can_reconstruct_query_key(&self) -> bool {
         DepKind::can_reconstruct_query_key(self)
     }
+
+    fn is_anon(&self) -> bool {
```
Comment (bjorn3) [inline compiler/rustc_middle/src/dep_graph/mod.rs:89]:
```suggestion
    #[inline]
    fn is_anon(&self) -> bool {
```

`can_reconstruct_query_key` also seems to be missing it.

**Comment by arora-aman** [comment]
> Triage: Is this ready for review? There's still [WIP] in the PR title.

@wesleywiser are we planning on merging this change? If so I can push a fixup.

**Comment by wesleywiser** [comment]
@arora-aman Don't we still need to do this? https://github.com/rust-lang/rust/issues/45408#issuecomment-365656548

I tried building locally and the incremental cache seems to be the same size before and after this change. I would assume we should see some kind of change if the nodes are no longer being allocated.

**Comment by oli-obk** [comment]
r? @wesleywiser 

I have no idea what this PR is about :laughing:

**Comment by tgnottingham** [comment]
Regarding the overall approach, we can't remove the anonymous nodes without compensating for their removal (most importantly, copying their dependencies up to the things that depend on them). I've made [comments](https://github.com/rust-lang/rust/issues/45408#issuecomment-723512275) on #45408 discussing some of the issues involved. TL;DR, I think the problem to be solved is a bit harder than it first appears, and solving it isn't necessarily a clear win.

Regarding the code, since I suspect the overall approach will change, my comments are probably totally irrelevant in the end. But here they are nonetheless.

* `with_task_impl` is supposed to return the allocated node index. If we're not allocating a node, it would be incorrect to say that we were by returning an index. And you wouldn't want to use `next_virtual_depnode_index`. That's only used when the dep graph is disabled (i.e. non-incremental mode unless some debug options are enabled). If we return that, it'll end up being the index of some other node in the graph most likely.

* That's all irrelevant though, because that code doesn't ever execute (I'm pretty certain anyway). Anonymous queries are executed using `with_anon_task`, which doesn't use `with_task_impl`. In fact, it's probably an error if `with_task_impl` is called with an anonymous node.

  Given that, it seems appropriate to have `finish_task_and_alloc_depnode` not return an `Option`, since the code isn't equipped to handle that case. Callers that create the closure can do the unwrapping of the `Option` returned by `alloc_node`. You could also just go back to unwrapping in `with_task_impl`, but it seems silly to suggest to clients that `finish_task_and_alloc_depnode` can really return `None`, when we're just going to panic if it does.

* The reason you aren't seeing a reduction in size is that anonymous queries use `intern_node`, not `alloc_node`. See `complete_anon_task`.

  That being said, I think it should be using `alloc_node`. `intern_node` is used when we think the graph might already contain the node. I don't think that's the case here. I don't think we should be executing the anonymous query if its node is already in the graph. Somewhere higher in the stack, we should have been using the query's cached result.

**Comment by wesleywiser** [comment]
Let's go ahead and close this as @arora-aman [suggested on Zulip](https://rust-lang.zulipchat.com/#narrow/stream/241847-t-compiler.2Fwg-incr-comp/topic/.2377070/near/216400647) since they're not going to have time to work on this more in the near future. 

@arora-aman if want to continue working on this, feel free to re-open at any point!

**Pull request closed without merging**